Post: dropping MySQL

lzap opened the pull request:
Do **not** comment on grammar or typos, please **checkout and push changes** into my branch. If there are many comments I will ask you to do so :-)

Thanks.

lzap committed:
MySQL upgrade instructions and blog post

ares commented:
reads well, :+1:, is there some timing expectation or can we merge as is right away? is the banner for 1.23 ready?

tbrisker commented on _posts/2019-06-28-dropping-support-for-mysql.md:
```diff
@@ -0,0 +1,47 @@
+
+It was not easy decision, we tried hard in the past to support both databases. If you search for [PRs containing MySQL](https://github.com/theforeman/foreman/pulls?page=3&q=is%3Apr+mysql) you'll find counless workarounds, fixes and endless discussions about how to solve problems in a way that doesn't break for PostgreSQL users. There is one common theme everywhere: most developers do not use MySQ ...
+
+We don't stop here, the next target is dropping SQLite3 forcing all developers to use PostgreSQL. It's the only reasonable way we can do in order to leverage all powerful features from the database system of our choice. However, again it's not easy. In the build process, we have a part that generates what's called APIPie bindings, a stack which provides API documentation and bindings metadata. An ...
```
I don't think this paragraph is relevant for our users. This is a more developer oriented issue.

lzap replied:
However is it a reason not to communicate this? My thinking it - I don't want to leave users just with simple "it will be gone", I want to be as much as transparent about our plans, goals and long term vision.

ekohl replied:
Perhaps you can link to a Discourse RFC that talks about the future?

lzap replied:
Dropped it.

tbrisker commented on _posts/2019-06-28-dropping-support-for-mysql.md:
```diff
@@ -0,0 +1,47 @@
+
+Along with the 1.23 release, we will provide official migration documentation, and mention it in our Release Notes as well. If you want to give the migration a test shot today, there's an existing [migration blog post](https://theforeman.org/2017/10/migrating-foreman-from-mysql-to-postgresql.html) written by Timo. We don't expect the official document be very much different. If you do such testin ...
+
+It was not easy decision, we tried hard in the past to support both databases. If you search for [PRs containing MySQL](https://github.com/theforeman/foreman/pulls?page=3&q=is%3Apr+mysql) you'll find counless workarounds, fixes and endless discussions about how to solve problems in a way that doesn't break for PostgreSQL users. There is one common theme everywhere: most developers do not use MySQ ...
```
Red Hat isn't the reason developers don't use MySQL, some RH developers actually do, and in any case this isn't the reason for moving away from PG. I would drop that sentence.

lzap replied:
Sure removing.

tbrisker commented on _posts/2019-06-28-dropping-support-for-mysql.md:
```diff
@@ -0,0 +1,47 @@
+
+This is not an extensive list you might think. Supporting two databases must not be that hard. Truth is, it's a challenge.
+
+Those two items on the list are both relational database systems with SQL standard, the problem is that very often behavior is different. Historically, some features were implemented differently, however it's pretty clear that PostgreSQL is now much closer to the RDBM "ideal" for both developers and operations, should I say. There are dozens of things that are just not right in MySQL and we need  ...
```
Each DB has it's own merits. We aren't moving to PG because it's "better" but rather because supporting just one has its benefits, and we already have plugins and other services that require it. PG also has weird problems we need to identify and fix, but having less databases to support will reduce the amount of such workarounds and will stop forcing us to always write db-agnostic code.

ekohl replied:
https://about.gitlab.com/2019/06/27/removing-mysql-support/ is a similar blog about the exact same topic. One thing stood out to me:

> As a side note – we don't use MySQL ourselves, and not doing so meant we weren't encountering issues BEFORE our users.

Both commercial vendors (Red Hat and ATIX) both only deploy on PostgreSQL. That means they care about proper support. MySQL always has been best effort by the community. The Foreman project wants to deliver high quality software and that's more than just 'it compiles, ship it'.

That's at least my take on this.

lzap replied:
I am rewriting the paragraph and dropping all my biased words about MySQL :-)

tbrisker reviewed (changes requested):
Fixed some grammar issues, other comments inline. Also, where is the banner for 1.23 added?

lzap commented:
Added new commit with the change, please make sure to squash on merge.

lzap commented:
Ping, I actually rebased once again and created new chapter 3.7 in the documentation which is based on Timos article but I added past experience with MySQL migrations we are seeing on Discourse. This is now ready.

tbrisker commented on manuals/nightly/index.md:
```diff
@@ -64,6 +64,8 @@ previous_version: "1.22"
 ## 3.6 Upgrade
 {%include manuals/{{ page.version }}/3.6_upgrade.md %}
+## 3.7 Migrating to PostgreSQL
+{%include manuals/{{ page.version }}/3.7_migration.md %}
```
forgot `git add`?

lzap replied:
Oh yes, pretty much :-)

tbrisker commented on _posts/2019-06-28-dropping-support-for-mysql.md:
```diff
@@ -0,0 +1,45 @@
+
+Along with the 1.23 release, we will provide official [migration documentation](/manuals/nightly/index.html#3.7MigratingtoPostgreSQL), and mention it in our Release Notes as well. If you do such testing, please come back to us with time estimation so we can tell our users in advance what to expect.
+
+It was not easy decision, we tried hard in the past to support both databases. If you search for [PRs containing MySQL](https://github.com/theforeman/foreman/pulls?page=3&q=is%3Apr+mysql) you'll find counless workarounds, fixes and endless discussions about how to solve problems in a way that doesn't break for PostgreSQL users. There is one common theme everywhere: most developers do not use MySQ ...
```
```suggestion
It was not easy decision, we tried hard in the past to support both databases. If you search for [PRs containing MySQL](https://github.com/theforeman/foreman/pulls?page=3&q=is%3Apr+mysql) you'll find countless workarounds, fixes and endless discussions about how to solve problems in a way that doesn't break for PostgreSQL users. There is one common theme everywhere: most developers do not regularly use MySQL both for development and for production testing. This is compounded by the fact that the two commercial products based on Foreman, Red Hat Satellite and Atix Orcharhino, only support PostgreSQL, and many of the Foreman developers are employed by them. As none of their customers use MySQL, there is less testing and usage of it leading to longer times until bugs are discovered and fixed. Also, one of the biggest and most popular plugins, Katello, does not support MySQL, so many community users are not using it either.
```

lzap replied:
Actually I dropped the whole paragraph, I just forgot to push properly.

tbrisker commented on _posts/2019-06-28-dropping-support-for-mysql.md:
```diff
@@ -0,0 +1,45 @@
+
+**Foreman 1.23 will be the last version with MySQL support.**
+
+The upcoming version 1.23 will also feature a warning UI strip in case it's running on MySQL. Don't worry, we will let you close the warning as soon as you see it, but we want to communicate this properly and in advance. We highly recommend to do the database migration to PostgreSQL during upgrade to version 1.23 because if there is a blocker bug that prevents from migrating to MySQL, there is st ...
```
The UI warning is still missing

lzap replied:
Going to push PR shortly.

lzap commented:
Squashed, rebased, added the missing file, renamed the filename to more appropriate date "1st of August".

lzap commented:
Here is the change that squashing hide for you to review, sorry about that:

```diff
--- a/_posts/2019-06-28-dropping-support-for-mysql.md
+++ b/_posts/2019-06-28-dropping-support-for-mysql.md
@@ -1,7 +1,7 @@
 ---
 layout: post
 title: Dropping support for MySQL
-date: 2019-06-28  9:34:21
+date: 2019-07-22 9:34:21
 author: Lukas Zapletal
 tags:
 - foreman
@@ -18,9 +18,9 @@ Foreman 1.22 currently requires one of the following databases:
 
 This is not an extensive list you might think. Supporting two databases must not be that hard. Truth is, it's a challenge.
 
-Those two items on the list are both relational database systems with SQL standard, the problem is that very often behavior is
 different. Historically, some features were implemented differently, however it's pretty clear that PostgreSQL is now much closer to the RDBM "ideal" for both developers and operations, should I say. There are dozens of things that are just not right in MySQL and we need to identify those problems, find a way to fix them and then carry extra bit of code in our codebase for life. This takes a lot of our time we could invest in features, bugfixing or performance improvements.
+Those two items on the list are both relational database systems with SQL standard, the problem is that very often behavior is different. Most of our development team use PostgreSQL and there are dozens of things that are different in MySQL. Therefore we need to identify those problems, find a way to fix them and then carry extra bit of code in our codebase for life. This takes a lot of our time we could invest in features, bugfixing or performance improvements.
 
-We have dicussed preferring PostgreSQL many times, but [the last time was different](https://community.theforeman.org/t/minimum-rdbm-requirements/14127) and we came to conclusion it's the right thing to do. [The plan](https://community.theforeman.org/t/rfc-dropping-support-for-mysql/14185) was then set and we are at the beginning of the journey where the first and really difficult step is on you, our users:
+We have dicussed preferring PostgreSQL many times, but [the last time was different](https://community.theforeman.org/t/minimum-rdbm-requirements/14127) and we came to conclusion that dropping MySQL the right thing to do. [The plan](https://community.theforeman.org/t/rfc-dropping-support-for-mysql/14185) was then set and we are at the beginning of the journey where the first and really difficult step is on you, our users:
 
 **Foreman 1.23 will be the last version with MySQL support.**
 
@@ -28,13 +28,11 @@ The upcoming version 1.23 will also feature a warning UI strip in case it's runn
 
 **Do not wait with the migration for Foreman 1.24.**
 
-Along with the 1.23 release, we will provide official migration documentation, and mention it in our Release Notes as well. If you want to give the migration a test shot today, there's an existing [migration blog post](https://theforeman.org/2017/10/migrating-foreman-from-mysql-to-postgresql.html) written by Timo. We don't expect the official document be very much different. If you do such testing, please come back to us with time estimation so we can tell our users in advance what to expect.
+Along with the 1.23 release, we will provide official [migration documentation](/manuals/nightly/index.html#3.7MigratingtoPostgreSQL), and mention it in our Release Notes as well. If you do such testing, please come back to us with time estimation so we can tell our users in advance what to expect.
 
-It was not easy decision, we tried hard in the past to support both databases. If you search for [PRs containing MySQL](https://github.com/theforeman/foreman/pulls?page=3&q=is%3Apr+mysql) you'll find counless workarounds, fixes and endless discussions about how to solve problems in a way that doesn't break for PostgreSQL users. There is one common theme everywhere: most developers do not use MySQL both for development and for production testing. This stems from the fact that Red Hat, a company which pays most of the users to do day-to-day development of Foreman, only have a product that is based on PostgreSQL. So we often try to reproduce things on PostgreSQL or blindly provide patches and let affected users to hotfix their instances. Also the biggest and most popular plugin called Katello does not support MySQL, so many community users are not using it at all. Although we haven't asked about it explicitly in the Foreman Community Survey yet, we believe that the majority of our users use PostgreSQL already.
+It was not easy decision, we tried hard in the past to support both databases. If you search for [PRs containing MySQL](https://github.com/theforeman/foreman/pulls?page=3&q=is%3Apr+mysql) you'll find counless workarounds, fixes and endless discussions about how to solve problems in a way that doesn't break for PostgreSQL users. There is one common theme everywhere: most developers do not use MySQL both for development and for production testing. This stems from the fact that Red Hat, a company which pays most of the users to do day-to-day development of Foreman, only have a product that is based on PostgreSQL. So we often try to reproduce things on PostgreSQL or blindly provide patches and let affected users to hotfix their instances. Also the biggest and most popular plugin called Katello does not support MySQL, so many community users are not using it at all.
 
-We don't stop here, the next target is dropping SQLite3 forcing all developers to use PostgreSQL. It's the only reasonable way we can do in order to leverage all powerful features from the database system of our choice. However, again it's not easy. In the build process, we have a part that generates what's called APIPie bindings, a stack which provides API documentation and bindings metadata. And this needs to boot up Rails, and Rails needs a database to run on and we all do this in a build chroot which does not provide environment for starting up daemon like PostgreSQL. We need to figure out how to workaround this problem, until then we still support SQLite3 for development but developers be ready to say goodbye to SQLite3 any new version.
-
-Let's evaluate after some time, because we already have ideas for many improvements we could do once PostgreSQL is the only supported RDBM. For example:
+We don't stop here, we already have ideas for more improvements we could do once PostgreSQL is the only supported RDBM. For example:
```

lzap commented:
Forgot to add your last minute change in the "It was not an easy decition" paragraph, now it should be ready for re-review. Sorry I was unable to revive changes from reflog, I screwed git hard today locally :-(

lzap commented:
Updated the text: 1.24 is the last version with MySQL. Added information about possible major version bump 2.0 (1.25) - it's not decided yet but we can communicate this early.

tbrisker commented on _includes/manuals/nightly/3.7_migration.md:
```diff
@@ -0,0 +1,185 @@
+management system. This chapter describes the process of migrating from MySQL
+to PostgreSQL.
+
+Full backup is absolute necessary before starting the process. We recommend
```
```suggestion
Full backup is absolutely necessary before starting the process. We recommend
```

tbrisker commented on _includes/manuals/nightly/3.7_migration.md:
```diff
@@ -0,0 +1,185 @@
+#### Prerequisites
+
+It's a good idea to delete unwanted records from the database. The smaller the
+database is, the faster will be the migration process. Typical offenders are
```
```suggestion
database is, the faster the migration process will be. Typical offenders are
```

tbrisker commented on _includes/manuals/nightly/3.7_migration.md:
```diff
@@ -0,0 +1,185 @@
+You can issue the MySQL command `SHOW FULL PROCESSLIST;` to show you all active
+connections to your MySQL database. Make sure, there are none.
+
+TODO: (WHY DO WE NEED THIS): If you are using the foreman-tasks plugin, you
```
Is this still needed? we should get this into the migration script if needed instead of manually patching from @timogoebel 's gist. cc: @adamruzicka

timogoebel replied:
dynflow uses tables that aren't managed by rails. That's why the migration script can't find a corresponding rails model and bails out.

adamruzicka replied:
What @timogoebel said, Dynflow uses Sequel to access the db and does its own migrations.

tbrisker replied:
Than why not just add the patch to the convert script? looking at the patch on the gist, it should be a noop if tasks isn't installed?

timogoebel replied:
Yeah, we can add it to core and ship it with a release.

tbrisker replied:
let's do it than and get it into 1.23 please, along with the post-migration sequence step.

lzap replied:
I think I saw a patch somewhere but can't find it. Can you guys point me on it if such a thing exists?

lzap replied:
Found it: https://github.com/theforeman/foreman/pull/6962

tbrisker commented on _includes/manuals/nightly/3.7_migration.md:
```diff
@@ -0,0 +1,185 @@
+down below after all instances are stopped.
+
+Foreman database must be cleaned up from any leftovers from uninstalled
+plugins. When a Foreman plugin is activated, database migration is perfored and
```
```suggestion
plugins. When a Foreman plugin is activated, database migration is performed and
```

tbrisker commented on _includes/manuals/nightly/3.7_migration.md:
```diff
@@ -0,0 +1,185 @@
+deactivated, no tables are ever dropped and the data is left in the database
+forever. It's good idea to review at least OS installed package log to find out
+if any foreman package was ever uninstalled. Unfortunately, there is no list of
+tables for each plugin.
```
I'm not sure how helpful this is - say they manage to find out what plugins were uninstalled, what then? we don't have instructions for cleaning up any leftovers and would need to help any users hitting issues because of that. It isn't just tables and indices, it could also be foreign keys, sti types and more. Perhaps it would be better to only warn that if any plugins have been uninstalled there may be issues and point to discourse for help or suggest reinstalling the plugins at least just for the migration step?

lzap replied:
Ok will do.

tbrisker commented on _includes/manuals/nightly/3.7_migration.md:
```diff
@@ -0,0 +1,185 @@
+
+#### Preparing the environment
+
+This blog post assumes that you already have the source and target databases
```
this is no longer a blog post

tbrisker commented on _includes/manuals/nightly/3.7_migration.md:
```diff
@@ -0,0 +1,185 @@
+
+The first step is to install required dependencies:
+
+    yum install -y foreman-postgresql
```
what about deb?

tbrisker commented on _includes/manuals/nightly/3.7_migration.md:
```diff
@@ -0,0 +1,185 @@
+    yum install -y foreman-postgresql
+
+Next, stop all your foreman instances and make sure, they stay stopped for
+good. On Red Hat compatible systems, you need to stop the `httpd` and (if
```
stay stopped during the migration

tbrisker commented on _includes/manuals/nightly/3.7_migration.md:
```diff
@@ -0,0 +1,185 @@
+    yum install -y foreman-postgresql
+
+Next, stop all your foreman instances and make sure, they stay stopped for
+good. On Red Hat compatible systems, you need to stop the `httpd` and (if
```
on debian?

lzap replied:
Not following.

tbrisker replied:
you gave the service name for rh systems, on debian it's apache2 iirc

tbrisker commented on _includes/manuals/nightly/3.7_migration.md:
```diff
@@ -0,0 +1,185 @@
+Next, stop all your foreman instances and make sure, they stay stopped for
+good. On Red Hat compatible systems, you need to stop the `httpd` and (if
+applicable) `foreman-tasks` service. There is a helper utility called
+`foreman-maintain` when installed to ease the process of shutting down
```
only on EL right now.

tbrisker commented on _includes/manuals/nightly/3.7_migration.md:
```diff
@@ -0,0 +1,185 @@
+starts these services again.
+
+You can issue the MySQL command `SHOW FULL PROCESSLIST;` to show you all active
+connections to your MySQL database. Make sure, there are none.
```
```suggestion
connections to your MySQL database. Make sure that there are none.
```

tbrisker commented on _includes/manuals/nightly/3.7_migration.md:
```diff
@@ -0,0 +1,185 @@
+#### Migrating
+
+Before proceeding migration, it's a good idea to permanently delete all audit
+records older than few days:
```
```suggestion
records older than few days, unless they are required for compliance reasons:
```

tbrisker commented on _includes/manuals/nightly/3.7_migration.md:
```diff
@@ -0,0 +1,185 @@
+
+#### Migrating
+
+Before proceeding migration, it's a good idea to permanently delete all audit
```
Add a heading - pre-migration cleanup? Also, perhaps it is simpler and faster to truncate all the reports and audits tables?

ohadlevy replied:
tbh I don't think we would like to ask users to delete their audits, only if they dont use it, I would just say it can speed things up, but recommend users to delete it.

lzap replied:
We do this via cron anyway both for reports and (now) for audits. I will update the heading.

lzap replied:
Marked the title as "optional" tho.

tbrisker commented on _includes/manuals/nightly/3.7_migration.md:
```diff
@@ -0,0 +1,185 @@
+`production.log`.
+
+To migrate the data from one database to the other, we will use a rake task.
+You might already be familiar to this task from upgrading Foreman to a newer
```
This isn't a task normally used in upgrades, i think we can drop this sentence.

tbrisker commented on _includes/manuals/nightly/3.7_migration.md:
```diff
@@ -0,0 +1,185 @@
+You might already be familiar to this task from upgrading Foreman to a newer
+version.
+
+Command `foreman-rake -T` lists all available tasks. We will be using
```
I think we can drop the -T sentence and combine the second one with the previous paragraph

tbrisker commented on _includes/manuals/nightly/3.7_migration.md:
````diff
@@ -0,0 +1,185 @@
+need to run all of Foreman's database migrations.
+
+```sh
+forman-rake db:migrate RAILS_ENV=development
````
Don't we also need to run `db:create` first?

ohadlevy replied:
exactly, really depending on how you configured pg, its users etc

tbrisker commented on _includes/manuals/nightly/3.7_migration.md:
```diff
@@ -0,0 +1,185 @@
+
+Postgresql uses sequences in order to store the next free primary key id for a
+table. These sequences do not get updated by the rake tasks and have to be
+adjusted manually. On MySQL, this is not needed as the concept of sequences is
```
Should we add this script to the migration script in case the target is pg?

timogoebel replied:
I think this would be great to have. Or another rake task that runs the commands on the database if it's mysql. Updating the sequences should be rather safe to do.

lzap replied:
Into which migration script, `prod2dev` is not supposed to do MySQL migrations.

tbrisker commented on _includes/manuals/nightly/3.7_migration.md:
````diff
@@ -0,0 +1,185 @@
+```sh
+su - postgres
+vim sequences.sql
+# Paste the script and save the file
````
Didn't we already save the file before?

lzap replied:
Replaced with https://github.com/theforeman/foreman/pull/6974

tbrisker commented on _includes/manuals/nightly/3.7_migration.md:
````diff
@@ -0,0 +1,185 @@
+
+```sh
+systemctl start httpd.service
+systemctl start foreman-tasks.service
````
doesn't httpd also start tasks now?

adamruzicka replied:
I'm not aware of it. Why would it do that?

ekohl replied:
No, it doesn't the installer ensures it's running. The current name of `foreman-tasks` is `dynflowd` btw. `foreman-tasks` is a legacy alias.

tbrisker replied:
Ok, ignore me, i was thinking of the dev environment which also starts dynflow.

tbrisker replied:
would still be good to change the service name though

lzap replied:
Changed to dynflowd then.

tbrisker commented on _posts/2017-10-31-migrating-foreman-from-mysql-to-postgresql.md:
```diff
@@ -7,6 +7,8 @@ tags:
 - foreman
 ---
 
+This content was moved to the [official Foreman manual](/manuals/nightly/index.html#3.7MigratingtoPostgreSQL).
```
```suggestion
*The [official Foreman manual](/manuals/1.23/index.html#3.7MigratingtoPostgreSQL) now contains the up to date and recommended migration procedure, please follow the instructions there.*
```

tbrisker commented on _posts/2019-08-01-dropping-support-for-mysql.md:
```diff
@@ -0,0 +1,90 @@
+PostgreSQL during upgrade to version 1.23 because if there is a blocker bug
+that prevents from migrating to MySQL, there is still a backup plan of holding
+off migration and staying on MySQL for one more release until we come up with
+versions 1.24 or 1.25 which will be PostgreSQL only. This is important:
```
```suggestion
a version which will be PostgreSQL only. This is important:
```

tbrisker commented on _posts/2019-08-01-dropping-support-for-mysql.md:
```diff
@@ -0,0 +1,90 @@
+**Do not wait with the migration for Foreman 1.24.**
+
+Along with the 1.23 release, we will provide official [migration
+documentation](/manuals/nightly/index.html#3.7MigratingtoPostgreSQL), and
```
```suggestion
documentation](/manuals/1.23/index.html#3.7MigratingtoPostgreSQL), and
```

lzap replied:
Why not to point to the most recent version when a random user finds the post after some time? Updated instructions might be quite helpful I am pretty sure we will run into issues.

tbrisker replied:
because we will not have that section in the manual once we drop mysql

lzap replied:
Why would we ever remove section named "Migration from MySQL to PostgreSQL" in the future? People are far behind with upgrading, we should keep it.

tbrisker replied:
because we don't want to carry around stuff in the manual that isn't used anymore, if people in the future do the upgrade they will still find the instructions in the 1.23/1.24 manual

tbrisker reviewed (changes requested):
See inline comments. @timogoebel - since the migration procedure is based on your blog post, could you please also give it a look?

tbrisker commented:
follow up thought - iiuc we explicitly skip the foreman tasks table, will the migration work for users with tasks? or will they lose all tasks information? cc @adamruzicka

adamruzicka commented:
From the top of my head, if we skip all `foreman_tasks_*` and `dynflow_*` tables they will lose:
- historic task data (can be exported before the migration)
- in-progress tasks (needs dynflow tables)
- recurring logic definitions
- tasks scheduled into the future (needs dynflow tables)
  - katello sync plans will have to be re-enabled to work again
  - recurring REX jobs will need to be wiped and set up again

tbrisker commented:
How should we migrate the tasks data then? I think historic data might be fine and there shouldn't be any running tasks ideally during the upgrade, but what about the others?
>     * katello sync plans will have to be re-enabled to work again
This is the only one we don't need to worry about at all since katello already requires postgres

ohadlevy commented on _includes/manuals/nightly/3.7_migration.md:
```diff
@@ -0,0 +1,185 @@
+#### Preparing the environment
+
+This blog post assumes that you already have the source and target databases
+set up. Make sure to setup the Postgresql database with UTF-8 enabled.
```
how does one do that? using the installer? imho you need to allow the user to copy-paste, or even write a small script to automate the migration.

lzap replied:
Yeah I also noticed that later on, I have removed this completely, everything is UTF-8 by default these days. Maybe @timogoebel remember why he added that?

ohadlevy commented on _includes/manuals/nightly/3.7_migration.md:
````diff
@@ -0,0 +1,185 @@
+```
+
+The foreman user that has access to your Postgresql database and is used for
+the conversion temporarily needs super user privileges.
````
how would one create the account in the first place? traditionally this has been done by the installer.

lzap replied:
You don't need super privileges, what you really need is grant to create tables which we handle in the installer. For the migration using super user is probably fine.

ohadlevy commented on _includes/manuals/nightly/3.7_migration.md:
```diff
@@ -0,0 +1,185 @@
+overwritten.
+
+You can configure the new database as Foreman's new development database by
+adding a new connection in `/etc/foreman/database.yml`.
```
tbh I would patch the script to read the definitions from elsewhere, we can provide a simple yaml file with source/destination instead of messing up with rails settings.

lzap replied:
Well, I am not keen doing that, I can test that once or twice on my environment and I don't want this to be rather for the worse.

ohadlevy reviewed (changes requested):
thanks for doing this, TBH IMHO we should just write a script that handles the migration (or as part of foreman maintain), it would be a lot easier to ask users to try it out vs giving them a doc.

adamruzicka commented:
> How should we migrate the tasks data then?

This should do the trick https://github.com/theforeman/foreman/pull/6962

lzap commented:
> TBH IMHO we should just write a script that handles the migration

Maybe it's just me but if I am migrating a production instance, I want to execute every one last of commands manually to see what I am doing (and taking notes). If others think this is a good idea I can write such thing however.

ohadlevy commented:
> > TBH IMHO we should just write a script that handles the migration
> 
> Maybe it's just me but if I am migrating a production instance, I want to execute every one last of commands manually to see what I am doing (and taking notes). If others think this is a good idea I can write such thing however.

my thought process that we need to make it easy and reliable, those people who care (like you) would want to see the procedure, how it works, and handle it step by step, others, would just want to get it running as easily as possible.

I think there is a lot of things that can go wrong in this process, having a script that is tested and improved over time, would make it more reliable than docs... my worry is that if its a hard procedure, people will simply not migrate...

btw you can still do a script that goes step by step and ask for confirmation of each step (or skip)...

lzap commented:
> btw you can still do a script that goes step by step and ask for confirmation of each step (or skip)...

Can I do that in foreman maintain @mbacovsky ?

lzap commented:
Forget about that @mbacovsky we can't use foreman-maintain as it is not on Debian.

lzap commented:
Added bunch of commits, squash during merge. There are still few open questions tho.

tbrisker commented on _includes/manuals/nightly/3.7_migration.md:
```diff
@@ -0,0 +1,171 @@
+
+### Migrating to PostgreSQL
+
+Starting from Foreman 1.25/2.0, MySQL is no longer a supported database
```
```suggestion
Starting from Foreman 1.25, MySQL is no longer a supported database
```
(at least until we agree on it)

tbrisker commented on _includes/manuals/nightly/3.7_migration.md:
```diff
@@ -0,0 +1,171 @@
+Next, stop all your foreman instances and make sure, they stay stopped during
+the migration. On Red Hat compatible systems, you need to stop the `httpd` and
+(if applicable) `foreman-tasks` service. On EL7, there is a helper utility
+called `foreman-maintain` when installed to ease the process of shutting down
```
```suggestion
called `foreman-maintain` that can be used to ease the process of shutting down
```

tbrisker commented on _includes/manuals/nightly/3.7_migration.md:
````diff
@@ -0,0 +1,171 @@
+```
+
+#### Pre-migrate cleanup (optional)
+
````
Perhaps first suggest truncating these table if they are not needed? that would be much faster then deleting possibly millions of records. Also, are there other tables we might suggest cleaning up, such as facts?

lzap replied:
I'd rather advertise the clean way of doing things. I am going to do add a note that truncate is an option.

tbrisker commented on _includes/manuals/nightly/3.7_migration.md:
```diff
@@ -0,0 +1,171 @@
+
+Next, stop all your foreman instances and make sure, they stay stopped during
+the migration. On Red Hat compatible systems, you need to stop the `httpd` and
+(if applicable) `foreman-tasks` service. On EL7, there is a helper utility
```
Still missing the service name for debian (apache2 iirc)

lzap replied:
Missed this one.

tbrisker commented on _includes/manuals/nightly/3.7_migration.md:
````diff
@@ -0,0 +1,171 @@
+migration of dynflow tables.
+
+```sh
+foreman-rake dynflow:migrate
````
Isn't this step done as part of the migration script now? Thinking of which, should we also run the regular migration task as part of it? (my main though is this is a very long process, the fewer manual steps the user has to do the better)

lzap replied:
I don't think so. More than that, it's not yet merged: https://github.com/theforeman/foreman/pull/6962

tbrisker replied:
That will be merged once @adamruzicka gets back to the comments. 
My thought is that these tasks should be idempotent, so we can call them as part of the prod2dev flow to make sure the target db is in the right state. if we do that, we can make these instruction simpler for the users - so less chance of missing a step causing the next step to fail.

tbrisker commented on _includes/manuals/nightly/3.7_migration.md:
````diff
@@ -0,0 +1,171 @@
+adjusted manually.
+
+```sh
+forman-rake sequence:reset RAILS_ENV=development
````
Let's call this as part of the convert script as well and not bother the user with another manual step

lzap replied:
I disagree, there is no such thing as the convert script. We do have `db:convert:prod2dev` which is a generic production to development convert script - your target database don't need to be PostgreSQL at all. I want to stick with least surprise, thus an explicit step is better. It's also better when calling for help - it's more obvious at which point something failed or what to do.

tbrisker replied:
Whenever prod2dev is called and the target db is postgres we would need to call this task, which is a noop anyways when the db isn't posgres. 
Users shouldn't need to care about database internals - they want to migrate from one db to the other. If this fails we'll be able to tell from the error what fails, and if they care about what the internals of the task does they can read it.

tbrisker commented on _includes/manuals/nightly/3.7_migration.md:
```diff
@@ -0,0 +1,171 @@
+To switch your Foreman instance to the new database, you need to change the
+production database configuration in `/etc/foreman/database.yml` to point to
+the development installation we configured earlier. You can either do this
+manually or re-run foreman-installer with the new parameters for Postgresql.
```
It's not clear what needs to be changed here if done manually or what are the new installer parameters for postgresql (and what parameters need to be reset to remove the mysql settings)

lzap replied:
Added bunch of instructions.

tbrisker commented on _includes/manuals/nightly/3.7_migration.md:
````diff
@@ -0,0 +1,171 @@
+Then you can finally start Foreman's services again.
+
+```sh
+systemctl start httpd.service
````
or apache2

lzap commented:
Pushed bunch of changes as new commits.

tbrisker reviewed (changes requested):
Almost there - I think the remaining issue is we disagree regarding having the convert script call other tasks to simplify the process. I would like users to have a simple path to migrate, the more complicated it is the more chance some of them will miss steps or just stick to 1.24 forever. 
The only other minor comment is that there is some inconsistent capitalization of PosgreSQL - some places use `Postgresql` instead.

lzap commented:
Pushing the PostgreSQL capitalization. I'd be fine with simplified workflow as long as we keep the current `prod2dev` task unchanged (because it's still quite useful task for development purposes) and introduce a different task "wrapper" that will perform everything what's needed.

tbrisker commented:
That works for me. lets make a `mysql_to_pg` task that wraps around these and does as much as possible on its own.

lzap commented:
Let me do this in https://github.com/theforeman/foreman/pull/6974 then. It's related.

lzap commented:
Integrated the new wrapper task.

tbrisker commented on _includes/manuals/nightly/3.7_migration.md:
````diff
@@ -0,0 +1,204 @@
+need to run all of Foreman's database migrations.
+
+```sh
+foreman-rake db:migrate RAILS_ENV=development
````
can we run this also inside the `mysql2pgsql` task?

lzap replied:
This can be dangerous - the db migration must be performed in the development Rails env. Now, the mysql2pgsql wrapper task today contains only conversion task which is Rails environment independent. However I am going to add another task which resets postgresql sequence and that already requires RAILS_ENV to be set correctly (production). If we merge the `db:migrate` here, there is a conflict in Rails environments.

It could be solved indeed but that can be confusing when adding more and more tasks into the wrapper one.

lzap commented:
Any other blockers @tbrisker other than the PRs to be merged?

tbrisker reviewed (approved):
This looks good now, but I prefer to finish merging the related PRs first to make sure we don't have a mismatch between the instructions here and what actually needs to be done.

tbrisker commented on _includes/manuals/nightly/3.7_migration.md:
````diff
@@ -0,0 +1,204 @@
+or tmux session.
+
+```sh
+foreman-rake db:convert:mysql2pgsql
````
```suggestion
foreman-rake db:convert:prod2dev
```

tbrisker commented on _posts/2019-08-01-dropping-support-for-mysql.md:
```diff
@@ -0,0 +1,90 @@
+administrators in case it's running on MySQL. Don't worry, we will let you
+close the warning as soon as you see it, but we want to communicate this
+properly and in advance. We highly recommend to do the database migration to
+PostgreSQL during upgrade to version 1.23 because if there is a blocker bug
```
Since we want them to use the updated script:
```suggestion
PostgreSQL right after upgrade to version 1.23 because if there is a blocker bug
```

tbrisker reviewed (changes requested):
Core is in, two small changes to this and also please copy the changes from nightly to the 1.23 manual as well.

tbrisker commented on _posts/2019-08-01-dropping-support-for-mysql.md:
```diff
@@ -0,0 +1,90 @@
+---
+layout: post
+title: Dropping support for MySQL
+date: 2019-08-01 9:34:21
```
Also this should be updated

lzap commented:
All done, ping me if something's wrong! Cheers.

tbrisker reviewed (approved):
Thanks @lzap !
